### ISSUE TITLE
Improve request body io-alike objects handling

### ISSUE THREAD

**Issue by ixti** [httprb/http#469]
As of #468 pipes will cause failure:

``` ruby
reader, writer = IO.pipe
writer.puts "abc"
writer.close

HTTP.post(url, :body => reader)
```

That is because `reader` is an IO that supports `rewind`, but it's illegal on pipes:

```
Errno::ESPIPE: Illegal seek
```

So, I think we should become more strict about what is allowed to be used as request body.

@httprb/core comments and thoughts are welcomed and appreciated!

**Comment by janko-m** [comment]
I think it would be nice to support pipes and other non-rewindable objects (another example would be [`Down::ChunkedIO`](https://github.com/janko-m/down#downchunkedio)), because HTTP.rb itself doesn't need the body to be rewindable. Is there any other existing or potential feature that would benefit from the constraint that request body needs to be rewindable?

I think that Webmock, which is the one that motivated this change, should replace the request body object with the accumulated string, once it reads the original request body. For me that seems like the ideal solution.

**Comment by ixti** [comment]
I agree that we should not limit body to be rewindable. We just need to raise proper error if response body can't be rewinded upon re-usage attempt.

About WebMock in fact I agree that we should improve integration. I was thinking that I can "better" patch request in webmock... And in fact after some thinking - I guess we can/should extract webmock integration into a gem under httprb - so that adapter will be "in-sync" with http gem itself.
